### DISCUSSION TITLE
QueryManager: Query Graph

### DISCUSSION THREAD

**Discussion by ls-1801** [nebulastream/nebulastream#328] [Ideas]
**P1:** The query manager's current implementation is complex. Over the years, its responsibilities have grown, and more functionality has been added to it. This is reflected in its public interface, which is anything but obvious and exposes many internal implementation details, which are proliferated across the entire NebulaStream codebase. For example, we find the query manager's status enum on the output of the query compiler, which tightly couples the query compiler with the query manager.

**P2**: Currently, we treat queries separately in our query manager implementation; this does not align with our vision of sharing partial queries wherever possible. Sharing Queries in the current query implementation has been an afterthought and requires stopping and replacing queries with a new shared query, in addition to new query states and special cases that need to be handled by the query manager.

**P3**: The current implementation of the query manager is not tested at all! It is a central piece of Nebulastream, and its test coverage has been neglected for years. This is mainly due to its large interface, which is very hard to test.

**G1**: Single Query Graph:

The QueryManager operates on a single graph and is unaware of individual queries. A different component can handle the construction of the query graph (offline). The query manager is solely responsible for ensuring the query graph is in the desired state and taking the necessary actions to move from the observed state into the desired state if it has been updated. This is based on the Kubernetes reconciliation pattern.

#### Individual Queries
```mermaid
flowchart TB
subgraph Query1
sk1[(sink-1)]
src1[\src-1/]

src1 --> p1((p1-1)) --> sk1
src1 --> p2((p2-1)) --> sk1
end

subgraph Query2
sk2[(sink-2)]
src2[\src-1/]

src2 --> p21((p1-1)) --> sk2
src2 --> p22((p2-2)) --> sk2
end
```
#### QueryGraph
```mermaid
flowchart TD
subgraph SharedQuery
sk1[(sink-1)]
sk2[(sink-2)]
src1[\src-1/]


src1 --> p2((p2-1)) --> sk1
src1 --> p1((p1-1)) --> sk1
p1((p1-1)) --> sk2
src1 --> p22((p2-2)) --> sk2
end
```
This will greatly improve the interface of the query manager as it limits the interface of the query manager to a single function: `setQueryGraph(QueryGraph desiredGraph)`, with the promise of eventually reaching the desired query graph state.

The optimizer can calculate operations like starting and stopping new queries by producing a new graph. This separates the `query optimization` from the `query deployment`. The interface allows other optimizations like query merging, query adaptations, and state migration of stateful operators.

**G2**: Verification. The QueryManager will be a non-trivial component, but it is worth it as it hides the complexity behind an easy-to-reason interface. The QueryManager is a central component of NebulaStream; even rare occurring faults are detrimental as the query manager is expected to be hit by a bulk of the system workload across many threads. 

The interface allows for easy testing. Test cases can be constructed by transitioning from Graph A to Graph B and verifying the anticipated actions are carried out correctly. For simpler test cases, we can employ `partial order reduction` to verify different thread schedules. Larger test cases composed of multiple anticipated actions rely on aggressive sanitization.

**Comment by alepping** [comment]
Hey @ls-1801, thanks for getting a cool conversation started on the QueryManager topic!
Imo, your idea sounds very interesting.
To me, an important thing to consider is that with this design, query sharing would be an invariant (design direction) of our system. That is, a key goal of our system that we strive for, is to support as many queries as possible, by merging queries (and sharing sources/sinks). This nudges the design system in a specific direction. Given that Ankit's reviewers often criticized the core motivation "who needs to run&share hundreds/thousands of queries", we should carefully consider whether we want to spend our limited resources towards that direction, since we will not have the resources to pursue other directions anymore.
I'm not opposed to the idea, I just want to make sure that we all carefully consider the direction of our system and agree on a bigger picture.

**Reply by ankitgit** [reply]
@alepping a small clarification. The reviewer's criticism is more about the lack of real-world queries that as a university entity we do not have access to unlike for instance Alibaba or Microsoft which tend to have large number of workloads. The paper ultimately got accepted with reviewers pointing to the novelty of the solution and that the solution is integrated into an actual system. If you want, we can discuss or I can share with you the review comments. In addition, there are a lot of papers in the context of MQO in streaming. However, no system supports this aspect natively. Therefore, we should consider this aspect to make our system stand out from SOTA. 

Another important point to consider here is that this refactoring is orthogonal to MQO as it only considers transitioning one query to another. This transition can also occur due to adaptive optimizations. For example, operator reordering due to the change in the workload or data characteristics.

**Reply by ankitgit** [reply]
Another point, MQO can also benefit when you only have 10s of long running queries ;)

**Reply by alepping** [reply]
Thanks for the clarification @ankitgit!

> The reviewer's criticism is more about the lack of real-world queries that as a university entity we do not have access to unlike for instance Alibaba or Microsoft which tend to have large number of workloads.

That is unfortunate criticism! But ultimately it is true, that we don't have access to a good (real-world-like) benchmark that we can use to demonstrate how MQO benefits real-world query processing. As I mentioned though, I'm not opposed to the idea. It should just be a careful consideration.

**Reply by ankitgit** [reply]
Another contribution I made as part of the ISQM paper was a synthetic query generator. It can be used to generate semantically equivalent queries. https://github.com/nebulastream/nebulastream-query-generator

**Comment by alepping** [comment]
> The optimizer can calculate operations like starting and stopping new queries by producing a new graph. This separates the query optimization from the query deployment.

> The query manager is solely responsible for ensuring the query graph is in the desired state and taking the necessary actions to move from the observed state into the desired state if it has been updated.

Just to make sure that I understood you correctly. So the idea is that the optimizer modifies the query plan(s) for its worker(s) and sends (or the worker(s) pull) an already calculated graph to the query manager. The job of the query manager then is to transition from the old query graph to the new one.

**Reply by keyseven123** [reply]
A follow-up question for you @ls-1801: Who is responsible for updating the graph? The gRPC thread or a worker thread?

**Comment by alepping** [comment]
> The QueryManager is a central component of NebulaStream; even rare occurring faults are detrimental as the query manager is expected to be hit by a bulk of the system workload across many threads.

So you are saying that we want to really make sure that there are as few faults as possible, because a fault can be detrimental to the query manager's performance and therefore to the entire system's performance?

**Reply by ls-1801** [reply]
It's not about a query failing. Query failure is always expected. This refers to all the diffeeent interleavings that could happen between updating the query graph under the query graph being traversed and executed by the worker threads.

Simplest problem would be that a pipeline is replaced with an adapted version and the query manager prematurely throws away the old pipeline, while it is still accessed by one of the worker threads.

**Reply by alepping** [reply]
> Simplest problem would be that a pipeline is replaced with an adapted version and the query manager prematurely throws away the old pipeline, while it is still accessed by one of the worker threads.

Thanks for the clarification. In general concurrent read/writes to the state of the query manager are problematic. I again assume/hope that concurrent read/write access to (specific types) of graphs is a well researched topic and there are algorithms (or even implementations) out there that we can rely on. There of course might be logic that is specific to how we process queries. In the best case we could fit our logic in existing/well-known patterns.

**Comment by alepping** [comment]
> Test cases can be constructed by transitioning from Graph A to Graph B and verifying the anticipated actions are carried out correctly.

I'd say this can even be proven formally. I assume there are algorithms that have been proven to be correct that enable our desired transitions. I imagine that our job is to model our transitions to conform to well known/proven algorithms, instead of a DIY solution. It of course still makes sense to test the actual implementation, which can still be faulty.

**Reply by keyseven123** [reply]
> I imagine that our job is to model our transitions to conform to well known/proven algorithms, instead of a DIY solution. It of course still makes sense to test the actual implementation, which can still be faulty.


Agree with this point!

**Comment by alepping** [comment]
Loose open questions:
What exactly happens when a query fails? So far I always imagined the role of the query manager to handle a failing query. What is the role of this new query manager in that case? Does it still expose something like an EmitFunction that can be used by the DataSource to notify the QueryManager that a specific source failed?
Given that the QueryManager does not know about individual queries anymore, how is such a failure communicated to other components of system? Is it communicated to the optimizer, which then calculates a new query graph, which the query manager then transitions to?
How is a failing query communicated to the user, who, most probably, would like to be informed on a per-query basis.
How do we handle per-query configuration? When can we merge to queries that are the same, but have different configurations? I assume we still apply pipelining, that is, the single-query-graph is split up into pipelines? Do you still envision a task queue for pipelines? We still need to think about the specifics of state transitions, e.g., when replacing one pipeline with another.
In case there are multiple workers. Do you envision a single optimizer that maintains multiple query graphs for each worker? Do you envision a single optimizer-query-graph that merges all worker-level-query-graphs? What about having multiple coordinators that potentially share workers? In general how does synchronization happen (worker-optimizer, optimizer-worker, worker-worker, optimizer-optimizer)?

**Comment by ankitgit** [comment]
Regarding **P2**, I like that QueryManager will expose only one public interface for handling reconfiguration of running plans. 

In this context, first, I think it is also important to define the overall set of reconfiguration types that we want to support such that a clear decision can be taken if only one interface is required for all cases (and if so how does it look) or if more are needed to handle different cases. For example:

1. Query sharing or adaptive optimization will result in the modification of running queries as you rightly mentioned above.
2. Terminating (stop/failure) or starting a query won't require any change to the target query.

Second, we also need to define who are the caller of the query manager that will call this interface. This will help in identifying if query manager is the right place to have this interface or something new is required. For example:   

1. Query modification due to the update in sharing opportunities or adaptive re-optimization will be request by the optimizer.
2. Query termination can be requested either by query (when a source is terminating or unrecoverable failure occurs) or from outside the system because the user wants to stop the query.
3. When deploying updates for a shared query plan across different workers, we need to ensure that these are changes are applied such that a tuple that arrived before the plan update will only be processed by older version of plans. Therefore, we will need coordination across nodes to apply these updates. This can be handled by using specialized markers like done in systems, such as, Chi, Rhino, Gloss.
